File: app/src/main/java/studio/mandysa/music/ui/base/BaseFragment.java
package studio.mandysa.music.ui.base;

import mandysax.fragment.Fragment;
import mandysax.lifecycle.ViewModel;
import studio.mandysa.music.ui.viewmodel.ShareViewModel;

public class BaseFragment extends Fragment {

    /**
     * @param viewmodel you viewmodel.class
     * @param <T> extends ViewModel
     * @return you viewmodel
     */
    public final <T extends ViewModel> T getViewModel(Class<T> viewmodel) {
        return getActivity().getViewModel(viewmodel);
    }

    /**
     * @return ShareViewModel
     */
    public final ShareViewModel getShareViewModel() {
        return getActivity().getShareViewModel();
    }

    /**
     * @return BaseActivity
     */
    @Override
    public BaseActivity getActivity() {
        return (BaseActivity) super.getActivity();
    }

}

File: app/src/main/java/studio/mandysa/music/ui/base/BaseFragment.kt
package studio.mandysa.music.ui.base

import android.view.LayoutInflater
import androidx.viewbinding.ViewBinding
import mandysax.fragment.Fragment
import mandysax.lifecycle.Lifecycle
import mandysax.lifecycle.ViewModel
import mandysax.lifecycle.ViewModelProviders
import kotlin.properties.ReadOnlyProperty
import kotlin.reflect.KProperty

open class BaseFragment : Fragment() {

    inline fun <reified VB : ViewBinding> bindView() =
        FragmentBindingDelegate(VB::class.java)

    class FragmentBindingDelegate<VB : ViewBinding>(
        private val clazz: Class<VB>
    ) : ReadOnlyProperty<Fragment, VB> {

        private var _binding: VB? = null

        override fun getValue(thisRef: Fragment, property: KProperty<*>): VB {
            if (_binding == null) {
                _binding = clazz.getMethod("inflate", LayoutInflater::class.java)
                    .invoke(null, thisRef.layoutInflater) as VB
                thisRef.viewLifecycleOwner.lifecycle.addObserver { state ->
                    if (state == Lifecycle.Event.ON_DESTROY)
                        _binding = null
                }
            }
            return _binding!!
        }
    }

    inline fun <reified VB : ViewModel> BaseFragment.viewModels() = lazy {
        return@lazy ViewModelProviders.of(
            viewLifecycleOwner
        )[VB::class.java]
    }

    inline fun <reified VB : ViewModel> BaseFragment.activityViewModels() = lazy {
        return@lazy ViewModelProviders.of(
            activity
        )[VB::class.java]
    }

}